### PR TITLE
Change Android template launchMode to singleTask

### PR DESCRIPTION
**Pull request by lukepighetti**
We were having issues with deeplinking relaunching activities/apps on huawei and samsung devices, switching to singleTask resolved this issue. Should we change the flutter template to use `singleTask`?

Since I'm not completely familiar with the downstream changes this might cause I would appreciate thoughtful review of this PR. This PR is intended to ask a question, and if this happens to be the answer, I wanted to make it easy to pull in.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
According to https://developer.android.com/guide/topics/manifest/activity-element#lmode

singleTask is listed as "Specialized launches (not recommended for general use)"

**Comment by blasten** [comment]
@jonahwilliams won. ditto.

**Comment by jonahwilliams** [comment]
@lukepighetti to make progress here, we'd need to figure out why switching to single task fixed the bug. because the android documentation doesn't recommend it, I wouldn't feel comfortable switching the template default until we have a better idea of why it resolved your issue. i.e. is this a bug in the flutter android embedding, or an issue with the android sdk, or even just a bug in these specific devices.

Would you be willing to file an issue describing the bug you were hitting, including any possible steps to reproduce?

**Comment by lukepighetti** [comment]
Yeah let’s start there and I will be happy to close this.

**Comment by niwatly** [comment]
The same issue here, maybe same as this.

https://github.com/nordnet/cordova-universal-links-plugin/issues/37

**Comment by lukepighetti** [comment]
Reopening this.

Again, I'm not an Android developer so I'd appreciate someone to backfill where I have gaps.

As I understand it, Flutter `.MainActivity` should always be a single activity unless an app has a unique use case, which it would be their responsibility to change it to `singleTop`.

This manifests itself when handling something like tapping on a push notification in Android. In `singleTop` mode it appears that this intent spawns a new instance of a Flutter app by spawning a new `.MainActivity`. Now, I expect there's a way to control this in the native Android code, but this appears to be the way it works out of the box for every push notification package I've tried yet (Firebase Messaging & OneSignal).

I've tried to find more documentation surrounding this but I haven't been able to find anything conclusive. Can anyone chime in and provide more context here?

**Comment by jonahwilliams** [comment]
@blasten could you follow up on this?

**Comment by blasten** [comment]
> As I understand it, Flutter .MainActivity should always be a single activity unless an app has a unique use case, which it would be their responsibility to change it to singleTop. That's because Flutter apps have their own navigation stack within the Flutter view and the idea of a Flutter app launching twice in a native stack is not something Flutter developers typically have as a use case.

That's right.

> This manifests itself when handling something like tapping on a push notification in Android. 

I see. We also have a number of Flutter apps (internally) that use `singleTask`

**Review by blasten — APPROVED** [review]
LGTM

File: packages/flutter_tools/templates/app/android.tmpl/app/src/main/AndroidManifest.xml.tmpl [hunk 1]
```diff
@@ -11,7 +11,7 @@
         android:icon="@mipmap/ic_launcher">
         <activity
             android:name=".MainActivity"
-            android:launchMode="singleTop"
+            android:launchMode="singleTask"
```
Comment (blasten) [inline packages/flutter_tools/templates/app/android.tmpl/app/src/main/AndroidManifest.xml.tmpl:14]: do you mind also changing the plugin example app template?

Comment (lukepighetti) [reply packages/flutter_tools/templates/app/android.tmpl/app/src/main/AndroidManifest.xml.tmpl:14]: This should be resolved in https://github.com/flutter/flutter/pull/56899/commits/7a4dcd31ac5cd3464eaa02186f6fd3f26a910b94

Comment (blasten) [reply packages/flutter_tools/templates/app/android.tmpl/app/src/main/AndroidManifest.xml.tmpl:14]: actually, there's no such a thing as a "plugin example app template". Thanks for updating the module though

**Comment by letsar** [comment]
By using singleTask by default, all Flutter apps would be vulnerable to this exploit: https://blog.dixitaditya.com/android-task-hijacking/amp/
I'm not a security expert neither a native android developer, but could we use singleInstance and set taskAffinity to an empty String instead, in order to avoid to be vulnerable by this exploit?

**Comment by lukepighetti** [comment]
I'm not qualified to comment on risk but here's a related link: https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-ren-chuangang.pdf

**Comment by zanderso** [comment]
@blasten does this PR need another look?

**Comment by jonahwilliams** [comment]
It looks like this needs to be rebased in order to pick up fixes to our CI config, and to resolve the merge conflict

**Comment by zanderso** [comment]
@lukepighetti Are you interested in continuing with this PR?

**Comment by christopherfujino** [comment]
closing this PR as outdated. Feel free to re-open if you want to pick this up again!

**Pull request closed without merging**